dkg/frost/frost.go

```diff
@@ -0,0 +1,756 @@
+	}
+
+	out := &dkg.KeyGenOutput{
+		Share:           sk,
+		OperatorPubKeys: operatorPubKeys,
+		ValidatorPK:     vk,
+		Threshold:       uint64(fr.threshold),
+	}
+	return out, nil
+}
+
+func (fr *FROST) processBlame() (*dkg.BlameOutput, error) {
+	for operatorID, msg := range fr.msgs[Blame] {
+		protocolMessage := &ProtocolMsg{}
+		if err := protocolMessage.Decode(msg.Message.Data); err != nil {
+			return nil, fmt.Errorf("failed to decode blame data")
+		}
+
+		var (
+			valid bool
+			err   error
+		)
+
+		switch protocolMessage.BlameMessage.Type {
+		case InvalidShare:
+			valid, err = fr.processBlameTypeInvalidShare(operatorID, protocolMessage.BlameMessage)
+			if err != nil {
+				return nil, err
+			}
+		case InconsistentMessage:
+			valid, err = fr.processBlameTypeInconsistentMessage(operatorID, protocolMessage.BlameMessage)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		blameMessageBytes, err := protocolMessage.BlameMessage.Encode()
+		if err != nil {
+			return nil, err
+		}
+
+		blameOutput := &dkg.BlameOutput{
+			Valid:        valid,
+			BlameMessage: blameMessageBytes,
+		}
+		return blameOutput, nil
+	}
+	return nil, nil
+}
+
+func (fr *FROST) processBlameTypeInvalidShare(operatorID uint32, blameMessage *BlameMessage) (bool /*valid*/, error) {
+	round1Message := &Round1Message{}
+	if err := json.Unmarshal(blameMessage.BlameData[0], round1Message); err != nil {
+		return false, err
+	}
+
+	verifiers := new(sharing.FeldmanVerifier)
+	for _, commitmentBytes := range round1Message.Commitment {
+		commitment, err := thisCurve.Point.FromAffineCompressed(commitmentBytes)
+		if err != nil {
+			return false, err
+		}
+		verifiers.Commitments = append(verifiers.Commitments, commitment)
+	}
+
+	blamerSessionSK := ecies.NewPrivateKeyFromBytes(blameMessage.BlamerSessionSk)
+	shareBytes, err := ecies.Decrypt(blamerSessionSK, round1Message.Shares[operatorID])
+	if err != nil {
+		return false, err
+	}
+
+	share := &sharing.ShamirShare{
+		Id:    operatorID,
+		Value: shareBytes,
+	}
+
+	if err := verifiers.Verify(share); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (fr *FROST) processBlameTypeInconsistentMessage(operatorID uint32, blameMessage *BlameMessage) (bool /*valid*/, error) {
+	if len(blameMessage.BlameData) != 2 {
+		return false, errors.New("invalid blame data")
+	}
+
+	var originalMessage, newMessage dkg.SignedMessage
+
+	if err := originalMessage.Decode(blameMessage.BlameData[0]); err != nil {
+		return false, err
+	}
+	if err := newMessage.Decode(blameMessage.BlameData[1]); err != nil {
+		return false, err
+	}
+
+	valid1 := originalMessage.Validate() == nil
+	valid2 := newMessage.Validate() == nil
+
+	return valid1 && valid2, nil
+}
+
+func (fr *FROST) canProceedRound1() bool {
+	if fr.currentRound != Preparation {
+		return false
+	}
+
+	for _, operatorID := range fr.operators {
+		protocolMessage := &ProtocolMsg{}
+
+		msg, ok := fr.msgs[Preparation][operatorID]
+		if ok {
```

Comment:
why is this tested here? Are we simply checking all operators sent a msg? if so please add some comments

Reply:
yes, we are simply checking if we have received all messages from the previous round. Will add comments and extract reusable code from this